### DISCUSSION TITLE
Memory leak when fetching data in useEffect with async/await

### DISCUSSION THREAD

**Discussion by bcreativex** [community/community#142736] [Programming Help]
### Body

I'm building a React component that fetches data from an API, but I'm getting a warning about memory leaks when navigating away from the component before the fetch completes. Here's my current implementation:
```
function UserProfile({ userId }) {
  const [userData, setUserData] = useState(null);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    async function fetchUserData() {
      const response = await fetch(/api/users/${userId});
      const data = await response.json();
      setUserData(data);
      setLoading(false);
    }
    
    fetchUserData();
  }, [userId]);

  // Rest of the component...
}
```
The console shows this warning:
Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application.

How can I properly handle cleanup to prevent these memory leaks?

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by Anikesh02** [comment]
This is a common issue when dealing with asynchronous operations in React components. The problem occurs when the component unmounts before the fetch operation completes. Here's how to fix it using an abort controller:
```
function UserProfile({ userId }) {
  const [userData, setUserData] = useState(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);

  useEffect(() => {
    const abortController = new AbortController();
    
    async function fetchUserData() {
      try {
        setLoading(true);
        setError(null);
        
        const response = await fetch(`/api/users/${userId}`, {
          signal: abortController.signal
        });
        
        if (!response.ok) {
          throw new Error(`HTTP error! status: ${response.status}`);
        }
        
        const data = await response.json();
        setUserData(data);
      } catch (error) {
        if (error.name === 'AbortError') {
          // Ignore abort errors
          return;
        }
        setError(error.message);
      } finally {
        setLoading(false);
      }
    }
    
    fetchUserData();

    return () => {
      abortController.abort();
    };
  }, [userId]);

  if (error) return <div>Error: {error}</div>;
  if (loading) return <div>Loading...</div>;
  if (!userData) return null;

  return (
    <div>
      <h1>{userData.name}</h1>
      {/* Rest of your component */}
    </div>
  );
}
```
This enhanced version includes:

- Proper error handling
- Loading states
- Cleanup on unmount
- Race condition prevention
- Conditional rendering based on state

**Comment by ebndev** [comment]
Please see [this comment](https://github.com/orgs/community/discussions/142737#discussioncomment-11078538) for reference to why this discussion was closed and locked. Thanks!
